File: app/src/main/java/com/example/customviewtestproject/customView/DashedCirclePainter.java
package com.example.customviewtestproject.customView;

import android.graphics.Canvas;
import android.graphics.DashPathEffect;
import android.graphics.Paint;
import android.graphics.RectF;

public class DashedCirclePainter implements Painter {

    private RectF dashedCircle;
    private Paint dashedCirclePaint;
    private int color;
    private int startAngle;
    private int finishAngle;
    private int width;
    private int height;
    private int strokeWidth;
    private int padding;

    DashedCirclePainter(int color, int startAngle, int finishAngle, int strokeWidth, int defaultPadding) {
        this.color = color;
        this.startAngle = startAngle;
        this.finishAngle = finishAngle;
        this.strokeWidth = strokeWidth;
        this.padding = defaultPadding;
        init();
    }

    private void init() {
        initDashedCirclePainter();
    }

    private void initDashedCirclePainter() {
        dashedCirclePaint = new Paint();
        dashedCirclePaint.setAntiAlias(true);
        dashedCirclePaint.setStrokeWidth(strokeWidth);
        dashedCirclePaint.setColor(color);
        dashedCirclePaint.setStyle(Paint.Style.STROKE);
        int dashWith = 5;
        int dashSpace = 8;
        dashedCirclePaint.setPathEffect(new DashPathEffect(new float[]{dashWith, dashSpace},
                dashSpace));
    }

    private void initDashedCircle() {
        dashedCircle = new RectF();
        float halfOffset = strokeWidth / 2f;
        dashedCircle.left = halfOffset;
        dashedCircle.top = halfOffset;
        dashedCircle.right = width - halfOffset;
        dashedCircle.bottom = height - halfOffset;
    }

    @Override
    public void draw(Canvas canvas) {
        canvas.drawArc(dashedCircle, startAngle, finishAngle, false, dashedCirclePaint);
    }

    public void setColor(int color) {
        this.color = color;
        dashedCirclePaint.setColor(color);
    }

    @Override
    public int getColor() {
        return color;
    }

    @Override
    public void onSizeChanged(int height, int width) {
        this.width = width - padding;
        this.height = height - padding;
        initDashedCircle();
    }
}

File: app/src/main/java/com/example/customviewtestproject/customView/DashedCirclePainter.kt
package com.example.customviewtestproject.customView

import android.graphics.Canvas
import android.graphics.DashPathEffect
import android.graphics.Paint
import android.graphics.RectF

class DashedCirclePainter(
    private var color: Int,
    private val startAngle: Int,
    private val finishAngle: Int,
    private val strokeWidth: Int,
    private val padding: Int
) : Painter {

    private var dashedCircle: RectF? = null
    private var dashedCirclePaint: Paint? = null
    private var width: Int = 0
    private var height: Int = 0

    init {
        dashedCirclePaint = Paint()
        dashedCirclePaint!!.isAntiAlias = true
        dashedCirclePaint!!.strokeWidth = strokeWidth.toFloat()
        dashedCirclePaint!!.color = color
        dashedCirclePaint!!.style = Paint.Style.STROKE
        val dashWith = 5
        val dashSpace = 8
        dashedCirclePaint!!.pathEffect = DashPathEffect(
            floatArrayOf(dashWith.toFloat(), dashSpace.toFloat()),
            dashSpace.toFloat()
        )
    }

    private fun initDashedCircle() {
        dashedCircle = RectF()
        val halfOffset = strokeWidth / 2f
        dashedCircle!!.left = halfOffset + strokeWidth
        dashedCircle!!.top = halfOffset + strokeWidth
        dashedCircle!!.right = width.toFloat() - halfOffset - strokeWidth.toFloat()
        dashedCircle!!.bottom = height.toFloat() - halfOffset - strokeWidth.toFloat()
    }

    override fun draw(canvas: Canvas) {
        canvas.drawArc(dashedCircle!!, startAngle.toFloat(), finishAngle.toFloat(), false, dashedCirclePaint!!)
    }

    override fun setColor(color: Int) {
        this.color = color
        dashedCirclePaint!!.color = color
    }

    override fun getColor(): Int {
        return color
    }

    override fun onSizeChanged(height: Int, width: Int) {
        this.width = width - padding
        this.height = height - padding
        initDashedCircle()
    }
}
